private/test/src/react/hooks.test.tsx

```diff
@@ -0,0 +1,252 @@
+    const value = Symbol(describeText);
+
+    resolveAfterWithValue(clientFn, 300, value);
+
+    afterEach(() => {
+      cleanup();
+    });
+
+    beforeEach(() => {
+      vi.clearAllMocks();
+    });
+
+    function render(): ReturnType<
+      typeof renderHook<AsyncResult<unknown>, TParams>
+    > {
+      return renderHook(hook, {
+        wrapper: App,
+        initialProps: initialParams,
+      });
+    }
+
+    it("initially returns loading AsyncState", () => {
+      const { result, unmount } = render();
+
+      expect(clientFn).toHaveBeenLastCalledWith(
+        initialParams,
+        expectedRequestOptions,
+      );
+      expect(clientFn).toHaveBeenCalledTimes(1);
+
+      expect(result.current.isLoading).toBe(true);
+      expect(result.current.data).toBeUndefined();
+      expect(result.current.error).toBeUndefined();
+
+      unmount();
+    });
+
+    it("returns data after loading", async () => {
+      const { result } = render();
+
+      await act(async () => await vi.advanceTimersByTimeAsync(300));
+      expect(clientFn).toHaveBeenCalledTimes(1);
+      expect(result.current.isLoading).toBe(false);
+      expect(result.current.data).toBe(value);
+      expect(result.current.error).toBeUndefined();
+    });
+
+    it("avoids unnecessary calls on Cerbos Client when rerendered with the values that match", async () => {
+      const { rerender, result } = render();
+      await act(async () => await vi.advanceTimersByTimeAsync(300));
+
+      rerender(
+        JSON.parse(JSON.stringify(initialParams)) as typeof initialParams,
+      );
+      expect(result.current.isLoading).toBe(false);
+      await act(async () => await vi.advanceTimersByTimeAsync(300));
+      expect(result.current.isLoading).toBe(false);
+      expect(clientFn).toHaveBeenCalledTimes(1);
+    });
+
+    it("calls Cerbos Client with new values when rerendered with different values", async () => {
+      const { rerender, result } = render();
+
+      await act(async () => await vi.advanceTimersByTimeAsync(300));
+
+      rerender(nextParams);
+      expect(clientFn).toHaveBeenLastCalledWith(
+        nextParams,
+        expectedRequestOptions,
+      );
+      expect(clientFn).toHaveBeenCalledTimes(2);
+      expect(result.current.isLoading).toBe(true);
+      await act(async () => await vi.advanceTimersByTimeAsync(300));
+      expect(result.current.isLoading).toBe(false);
+      expect(result.current.data).toBe(value);
+      expect(result.current.error).toBeUndefined();
+    });
+
+    it("aborts request when unmounted", () => {
+      let abortSignal: AbortSignal | undefined = undefined;
+
+      clientFn.mockImplementation((_, opt: RequestOptions) => {
+        abortSignal = opt.signal;
+      });
+
+      const { unmount } = render();
+
+      expect(abortSignal).toBeInstanceOf(AbortSignal);
+      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      expect(abortSignal!.aborted).toBe(false);
```

Comment:
Minor but here and below you could probably avoid the non-null assertion because the test will still fail if it is undefined:

```suggestion
      expect(abortSignal?.aborted).toBe(false);
```

Reply:
I thought the same but `no-unsafe-assignment` rule is not happy about that (no configuration options). I chose this one to suppress

Reply:
Ah, ok... weird!